### PR TITLE
[3.12] GH-106897: Add `RERAISE` event to `sys.monitoring`. (GH-107291)

### PR DESCRIPTION
**Pull request by markshannon**
Manual backport of GH-107291.

@Yhg1s as release manager.

Fixes an omission in PEP 669, resulting in an imbalance between raise and unwind/handle events for exceptions.

* Issue: gh-106897

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2023-07-26-12-18-10.gh-issue-106897.EsGurc.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Add a ``RERAISE`` event to ``sys.monitoring``, which occurs when an
+exception is reraise, either explicitly by a plain ``raise`` statement, or
```
Comment (Yhg1s) [inline Misc/NEWS.d/next/Core and Builtins/2023-07-26-12-18-10.gh-issue-106897.EsGurc.rst]:
```suggestion
exception is reraised, either explicitly by a plain ``raise`` statement, or
```

**Review by Yhg1s — APPROVED** [review]

**Comment by markshannon** [comment]
The ABI checking tool claims that the ABI has changed.
Why is the internal layout of `PyInterpreterState` part of the ABI?

**Comment by markshannon** [comment]
The ABI file diff is a quite a lot larger than I would have liked, but it seems to be OK.
@Yhg1s are you OK with merging the updated ABI file?

**Comment by markshannon** [comment]
I've waded through the `Doc/data/python3.12.abi` diff, and the changes are as expected, plus a lot of line number changes and some reordering.

**Pull request merged**